velox/dwio/common/DirectBufferedInput.cpp

```diff
@@ -29,6 +29,19 @@ using cache::CoalescedLoad;
 using cache::ScanTracker;
 using cache::TrackingId;
 
+DirectBufferedInput::~DirectBufferedInput() {
+  for (auto& load : coalescedLoads_) {
+    if (load->state() == cache::CoalescedLoad::State::kLoading) {
```

Comment:
@FelixYBW Binwei has a good point suggesting that we may need to add a close() method.

Reply:
+1. 
@FelixYBW I assume we only run into this problem in Gluten but not Prestissimo as Prestissimo will keep the finished task until all the pending reference drops. If we don't have such tasks (called Zombie Task in Prestissimo) on server shutdown, then there will be no memory leaks then?

Reply:
I don't think Prestissimo can avoid the issue. Here is the class reference chart. The problem here is DirectCoalescedLoad just link to memory pool, instead of refer to it by shared_ptr. So the memory pool's destructor either triggered by Task or MemoryManager can't wait until the DirectCoalescedLoad owned by Executor to release the memory, it can't release either because the pool is already destructed. 

So another possible fix solution is to use shared_ptr in DirectCoalesceLoad to hold the pool. But I don't think it's a reasonable way because which means even the task is already ended, some resource still keep live.

![image](https://www.plantuml.com/plantuml/png/VP7BJiCm44NtynMpGX0fVr1LvS42YGg4s4U3FOKgTXpvmKE5VyTEr2QE31idcPbxRkvuupnP7xG2eSWvlDbR5lugx7PiMTvrVV1m0CJtM3KMXI75Jh2yDoHT4njFJ9BjG-_tfdDencS3Fv01h-Uw0NYfb16Kfb6pRbxYLnlBMzR6VZqQeywmZ-T4ofqIOZvTeJwz9ybHD4qSJg3RJnOXfXaP9ELvuITHnUCYeZBGbZfwPGkZrBsHPTdqtgucd_NG-c1gbmsSQC_aslyXAKE-sxfUp-8076DWLTV5DKFXCeVXZ0jGIhyqoo6hxqWfq06oDIMpEIMbNjQAxjKV-0rtCcZr2m00)

Reply:
Or we can also add another taskholder to here.  But like #8198 I'd think adding close() is more reasonable way.

https://github.com/facebookincubator/velox/blob/322f30949d19077acb80ef7769f81e5307280989/velox/dwio/common/DirectBufferedInput.cpp#L186C9-L186C9

Reply:
Updated the code to use close function in the call chain.

Reply:
> I don't think Prestissimo can avoid the issue. Here is the class reference chart. The problem here is DirectCoalescedLoad just link to memory pool, instead of refer to it by shared_ptr. So the memory pool's destructor either triggered by Task or MemoryManager can't wait until the DirectCoalescedLoad owned by Executor to release the memory, it can't release either because the pool is already destructed.
> 
> So another possible fix is to use shared_ptr in DirectCoalesceLoad to hold the pool. But I don't think it's a reasonable way because which means even the task is already ended, some resource still keep live.
> 
> ![image](https://camo.githubusercontent.com/cfc2c05d9b8aa9da7ca7d2bdb05853a8cdae0c9c7d5fbf5fd9b8443d6a598ebe/68747470733a2f2f7777772e706c616e74756d6c2e636f6d2f706c616e74756d6c2f706e672f565037424a69436d34344e74796e4d70475830665672314c765334325947673473345533464f4b67545870766d4b453556795445723251453331696463506278526b767575706e5037784732655357766c446252356c7567783750694d5476725656316d30434a744d334b4d584937354a683279446f4854346e6a464a39426a472d5f74666444656e63533346763031682d5577304e59666231364b66623670526278594c6e6c424d7a5236565a71516579776d5a2d54346f6671494f5a7654654a777a39796248443471534a6733524a6e4f586658615039454c76754954486e554359655a4247625a667750476b5a724273485054647174677563645f4e472d633167626d735351435f61736c7958414b452d73786655702d3830373644574c545635444b4658436556585a306a47496879716f6f3668787157667130366f44494d7045494d624e6a5141786a4b562d30727443635a72326d3030)

@FelixYBW I see. DirectCoalescedLoad holds on raw reference to the child memory pool. If we change this to a shared pointer to hold it, then we might run into leaked memory pool as the child memory pool holds reference on task memory pool but task memory pool doesn't hold reference on task. Probably it is good to have close to wait for the pending loads to complete. Thanks!